lib/concurrent/executor/executor.rb

```diff
@@ -5,12 +5,75 @@
 module Concurrent
 
   module Executor
+    class AtExitAutoTermination
+      def initialize(enabled = true)
+        @mutex     = Mutex.new
+        @executors = Set.new
+        @enabled   = enabled
+        @installed = false
+      end
+
+      def add(executor)
+        @mutex.synchronize { @executors.add executor }
+      end
+
+      def remove(executor)
+        @mutex.synchronize { @executors.delete executor }
+      end
+
+      def enabled?
+        @mutex.synchronize { @enabled }
+      end
+
+      def enabled=(value)
+        @mutex.synchronize { @enabled = value }
+      end
+
+      def terminate
+        executors = @mutex.synchronize { @executors.to_a if @enabled }
+        if executors
+          executors.each(&:kill) # TODO be gentle first
+          executors.each(&:wait_for_termination)
+          executors.clear
+          true
+        else
+          false
+        end
+      end
+
+      def install
+        @installed ||= begin
+          at_exit { terminate }
+          true
+        end
+        self
+      end
+
+      def executors
+        @mutex.synchronize { @executors.to_a }
+      end
+    end
+
+    AT_EXIT_AUTO_TERMINATION = AtExitAutoTermination.new.install
```

Comment:
I also do not like the added complexity. However there is a valid case where the pools would be created periodically (maybe little obscure but I think still correct usage): long running background processor of CPU-heavy tasks of different nature, the processor can decide to create differently configured pools based on the task it is given to improve performance. The leak could cause problems if it runs long enough. For that reason I think we can justify the complexity exchanging it for fixed leak. Please look at the last commit, I've tried to reduce the complexity there https://github.com/ruby-concurrency/concurrent-ruby/commit/9db56e4a519d2f4a9b74af09c65076f305d533c6 by extracting the register to be a general `AtExit` manager.